display_list/geometry/dl_region.cc

```diff
@@ -60,189 +556,116 @@ std::vector<SkIRect> DlRegion::getRects(bool deband) const {
 
-  spans.push_back({left, right});
+bool DlRegion::isComplex() const {
+  return lines_.size() > 1 ||
+         (lines_.size() == 1 &&
+          span_buffer_.getChunkSize(lines_.front().chunk_handle) > 1);
 }
 
-bool DlRegion::SpanLine::spansEqual(const SpanLine& l2) const {
-  SpanVec& spans = *this->spans;
-  SpanVec& otherSpans = *l2.spans;
-  FML_DCHECK(this != &l2);
-
-  if (spans.size() != otherSpans.size()) {
+bool DlRegion::intersects(const SkIRect& rect) const {
+  if (isEmpty()) {
     return false;
   }
-  return memcmp(spans.data(), otherSpans.data(), spans.size() * sizeof(Span)) ==
-         0;
-}
-
-void DlRegion::insertLine(size_t position, SpanLine line) {
-  lines_.insert(lines_.begin() + position, line);
-}
 
-DlRegion::LineVec::iterator DlRegion::removeLine(
-    DlRegion::LineVec::iterator line) {
-  spanvec_pool_.push_back(line->spans);
-  return lines_.erase(line);
-}
+  auto bounds_intersect = SkIRect::Intersects(bounds_, rect);
 
-DlRegion::SpanLine DlRegion::makeLine(int32_t top,
-                                      int32_t bottom,
-                                      int32_t spanLeft,
-                                      int32_t spanRight) {
-  SpanVec* span_vec;
-  if (!spanvec_pool_.empty()) {
-    span_vec = spanvec_pool_.back();
-    spanvec_pool_.pop_back();
-    span_vec->clear();
-  } else {
-    span_vec = new SpanVec();
+  if (!isComplex()) {
+    return bounds_intersect;
   }
-  span_vec->push_back({spanLeft, spanRight});
-  return {top, bottom, span_vec};
-}
 
-DlRegion::SpanLine DlRegion::makeLine(int32_t top,
-                                      int32_t bottom,
-                                      const SpanVec& spans) {
-  SpanVec* span_vec;
-  if (!spanvec_pool_.empty()) {
-    span_vec = spanvec_pool_.back();
-    spanvec_pool_.pop_back();
-  } else {
-    span_vec = new SpanVec();
+  if (!bounds_intersect) {
+    return false;
   }
-  *span_vec = spans;
-  return {top, bottom, span_vec};
-}
-
-void DlRegion::addRects(std::vector<SkIRect>&& rects) {
-  std::sort(rects.begin(), rects.end(), [](const SkIRect& a, const SkIRect& b) {
-    // Sort the rectangles by Y axis. Because the rectangles have varying
-    // height, they are added to span lines in non-deterministic order and thus
-    // it makes no difference if they are also sorted by the X axis.
-    return a.top() < b.top();
-  });
 
-  size_t start_index = 0;
+  auto it = std::upper_bound(
+      lines_.begin(), lines_.end(), rect.fTop,
+      [](int32_t i, const SpanLine& line) { return i < line.bottom; });
 
-  size_t dirty_start = std::numeric_limits<size_t>::max();
-  size_t dirty_end = 0;
+  while (it != lines_.end() && it->top < rect.fBottom) {
+    FML_DCHECK(rect.fTop < it->bottom || it->top < rect.fBottom);
+    const Span *begin, *end;
+    span_buffer_.getSpans(it->chunk_handle, begin, end);
+    while (begin != end && begin->left < rect.fRight) {
+      if (begin->right > rect.fLeft && begin->left < rect.fRight) {
```

Comment:
You've checked `begin->left < rect.fRight` twice here.

Reply:
Indeed. Missed that.